pkg/kubectl/resize.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubectl
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+)
+
+// ResizePrecondition describes a condition that must be true for the resize to take place
+// If CurrentSize == -1, it is ignored.
+// If CurrentResourceVersion is the empty string, it is ignored.
+// Otherwise they must equal the values in the replication controller for it to be valid.
+type ResizePrecondition struct {
+	Size            int
+	ResourceVersion string
+}
+
+type PreconditionError struct {
+	Precondition  string
+	ExpectedValue string
+	ActualValue   string
+}
+
+func (pe *PreconditionError) Error() string {
+	return fmt.Sprintf("Expected %s to be %s, was %s", pe.Precondition, pe.ExpectedValue, pe.ActualValue)
+}
+
+// Validate ensures that the preconditions match.  Returns nil if they are valid, an error otherwise
+func (precondition *ResizePrecondition) Validate(controller *api.ReplicationController) error {
+	if precondition.Size != -1 && controller.Spec.Replicas != precondition.Size {
+		return &PreconditionError{"replicas", strconv.Itoa(precondition.Size), strconv.Itoa(controller.Spec.Replicas)}
+	}
+	if precondition.ResourceVersion != "" && controller.ResourceVersion != precondition.ResourceVersion {
```

Comment:
Hrm, I actually think that there are two cases, there are two conditions.

1) first being that the user's external constraints were violated, this is permanent in the absence of interaction with the user.

2) a transient race between read and write, and the user can immediately retry the same command and believe that it may succeed (or they'll get an error of type #1)